orc8r/cloud/go/services/analytics/analyzer.go

```diff
@@ -0,0 +1,182 @@
+	metricsConfig, err := config.GetServiceConfig(orc8r.ModuleName, metricsd.ServiceName)
+	if err != nil {
+		glog.Fatalf("Could not retrieve metricsd configuration: %s", err)
+	}
+	promClient, err := promAPI.NewClient(promAPI.Config{Address: metricsConfig.MustGetString(metricsd.PrometheusQueryAddress)})
+	// todo - for testing, will clean it up
+	//promClient, err := promAPI.NewClient(promAPI.Config{Address: "http://192.168.0.124:9090"})
+	if err != nil {
+		glog.Fatalf("Error creating prometheus client: %s", promClient)
+	}
+	return v1.NewAPI(promClient)
+}
+
+//NewPrometheusAnalyzer contructs a new prometheus analyzer instance
+func NewPrometheusAnalyzer(config *Config, prometheusClient v1.API, calculations []calculations.Calculation, exporter Exporter) Analyzer {
+	cronJob := cron.New()
+	return &PrometheusAnalyzer{
+		Config:           config,
+		Cron:             cronJob,
+		PrometheusClient: prometheusClient,
+		Calculations:     calculations,
+		Exporter:         exporter,
+	}
+}
+
+//Schedule method takes in a schedule string in cron format
+//and schedules the analyze job to be run at that schedule
+func (a *PrometheusAnalyzer) Schedule() error {
+	glog.Infof("Analytics job schedule %s", a.Config.AnalysisSchedule)
+
+	a.Cron = cron.New()
+	_, err := a.Cron.AddFunc(a.Config.AnalysisSchedule, a.LocalAnalyze)
+	if err != nil {
+		glog.Infof("error scheduling the local analytics function %v", err)
+		return err
+	}
+
+	_, err = a.Cron.AddFunc(a.Config.AnalysisSchedule, a.RemoteAnalyze)
+	if err != nil {
+		glog.Infof("error scheduling the remote analytics function %v", err)
+		return err
+	}
+	return nil
+}
+
+//LocalAnalyze method runs through all individual calculators
+//and exports the metrics if exporter is set
+func (a *PrometheusAnalyzer) LocalAnalyze() {
+	glog.V(1).Info("Running Local Analyze")
+	for _, calc := range a.Calculations {
+		results, err := calc.Calculate(a.PrometheusClient)
+		if err != nil {
+			glog.Errorf("Error calculating metric: %s", err)
+			continue
+		}
+		if a.Exporter == nil {
+			continue
+		}
+		for _, res := range results {
+			err = a.Exporter.Export(res, http.DefaultClient)
+			if err != nil {
+				glog.Errorf("Error exporting result: %v", err)
+			} else {
+				glog.Infof("Exported %s, %s, %f", res.MetricName, res.Labels, res.Value)
+			}
+		}
+	}
+}
+
+//RemoteAnalyze methods runs through remote collectors
+// and exports their metrics
+func (a *PrometheusAnalyzer) RemoteAnalyze() {
+	glog.V(1).Info("Running Remote Analyze")
+	collectorClients, err := getRemoteCollectors()
+	if err != nil {
+		glog.Infof("err %v failed to get remote collectors", err)
+		return
+	}
+	for _, c := range collectorClients {
+		collectResp, err := c.Collect(context.Background(), &protos.CollectRequest{})
+		if err != nil {
+			glog.Infof("err %v attempting to collect from service", err)
+			continue
+		}
+		if collectResp != nil {
+			for _, res := range collectResp.GetResult() {
+				if a.Exporter == nil {
```

Comment:
Can we just check this before the entire loop? If exporter is nil then we'll never do anything with the results so might as well just skip the loop if that's the case.

Reply:
I will add a comment here, we might still want to run through the loop in case the module level analytics services want to simply register the results into prometheus.